mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -270,21 +290,179 @@ impl NexusChild {
         Ok(self.name.clone())
     }
 
-    /// Register a key
-    pub(crate) async fn resv_register(
+    /// Register an NVMe reservation, specifying a new key
+    async fn resv_register(
         &self,
+        hdl: &dyn BlockDeviceHandle,
         new_key: u64,
+    ) -> Result<(), CoreError> {
+        hdl.nvme_resv_register(
+            0,
+            new_key,
+            nvme_reservation_register_action::REGISTER_KEY,
+            nvme_reservation_register_cptpl::NO_CHANGES,
+        )
+        .await?;
+        info!(
+            "{}: registered key {:0x}h on child {}",
+            self.parent, new_key, self.name
+        );
+        Ok(())
+    }
+
+    /// Acquire an NVMe reservation
+    async fn resv_acquire(
+        &self,
+        hdl: &dyn BlockDeviceHandle,
+        current_key: u64,
+        preempt_key: u64,
+        acquire_action: u8,
+        resv_type: u8,
     ) -> Result<(), ChildError> {
-        let hdl = self.get_io_handle().context(HandleOpen {})?;
-        if hdl.get_device().driver_name() == "nvme" {
-            info!(
-                "{}: registering key {:x}h on child {}...",
-                self.parent, new_key, self.name
+        if let Err(e) = hdl
+            .nvme_resv_acquire(
+                current_key,
+                preempt_key,
+                acquire_action,
+                resv_type,
+            )
+            .await
+        {
+            return Err(ChildError::ResvAcquire {
+                source: e,
+            });
+        }
+        info!(
+            "{}: acquired reservation type {:x}h, action {:x}h, current key {:0x}h, preempt key {:0x}h on child {}",
+            self.parent, resv_type, acquire_action, current_key, preempt_key, self.name
+        );
+        Ok(())
+    }
+
+    /// Get NVMe reservation report
+    /// Returns: (key, host id) of write exclusive reservation holder
+    async fn resv_report(
+        &self,
+        hdl: &dyn BlockDeviceHandle,
+    ) -> Result<Option<(u64, [u8; 16])>, ChildError> {
+        let mut buffer = hdl.dma_malloc(4096).context(HandleDmaMalloc {})?;
+        if let Err(e) = hdl.nvme_resv_report(1, &mut buffer).await {
```

Comment:
Would it be possible to wrap reservation data returned by 'nvme_resv_report' by a high-level data structure, to hide low-level manipulations with reservation info, like:
'buffer.as_slice().split_at(std::mem::size_of::<spdk_nvme_reservation_status_extended_data'),
ad expose only a well-defined API to access reservation info (like status, ) via this structure ?
That would let us keep NexusChild high-level enough from SPDK API and interact with the underlying block device via well-defined API

Reply:
I agree that this function is a little low-level for `NexusChild` but it's only used internally and partly done that way for expediency. If I were to move the various manipulations to `nvme_resv_report()`, then it'd probably return something like a tuple of `Options` of `spdk_nvme_reservation_status_extended_data` and and array of `spdk_nvme_registered_ctrlr_extended_data`. Is that what you had in mind?

Reply:
I was suggesting wrapping reservation info into some high-level structure(s), to let API users see only that placeholder.
If that takes longer to properly wrap it or it ends up in lots of unnecessary Options - then we can leave it as it is.

Reply:
I've opted to leave it alone for now.